test-app/src/main/java/org/readium/r2/testapp/MediaService.kt

```diff
@@ -0,0 +1,136 @@
+package org.readium.r2.testapp
+
+import android.app.PendingIntent
+import android.content.Intent
+import android.os.Build
+import android.os.IBinder
+import androidx.lifecycle.lifecycleScope
+import androidx.media2.session.MediaSession
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.*
+import org.readium.r2.navigator.ExperimentalAudiobook
+import org.readium.r2.navigator.media2.MediaNavigator
+import org.readium.r2.navigator.media2.MediaSessionNavigator
+import org.readium.r2.shared.util.Try
+import org.readium.r2.testapp.bookshelf.BookRepository
+import org.readium.r2.testapp.db.BookDatabase
+import org.readium.r2.testapp.reader.ReaderContract
+import org.readium.r2.testapp.utils.LifecycleMediaSessionService
+import timber.log.Timber
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.ExperimentalTime
+
+@OptIn(ExperimentalTime::class, ExperimentalAudiobook::class, ExperimentalCoroutinesApi::class)
+class MediaService : LifecycleMediaSessionService() {
+
+    inner class Binder : android.os.Binder() {
+
+        private val books by lazy {
+            BookRepository(BookDatabase.getDatabase(this@MediaService).booksDao())
+        }
+
+        private var currentBookId: Long? = null
+
+        private var saveLocationJob: Job? = null
+
+        var mediaSession: MediaSession? = null
+
+        var mediaNavigator: MediaSessionNavigator? = null
+
+        suspend fun openPublication(arguments: ReaderContract.Input): Try<Unit, MediaNavigator.Exception> {
+            closePublication()
+            return MediaSessionNavigator.create(
+                this@MediaService,
+                arguments.publication,
+                arguments.initialLocator
+            ).map {
+                bindNavigator(it, arguments)
+            }
+        }
+
+        fun closePublication() {
+            mediaSession?.close()
+            mediaSession = null
+            saveLocationJob?.cancel()
+            saveLocationJob = null
+            mediaNavigator?.close()
+            mediaNavigator = null
+            currentBookId = null
+        }
+
+        @OptIn(FlowPreview::class)
+        private fun bindNavigator(navigator: MediaSessionNavigator, arguments: ReaderContract.Input) {
+            val activityIntent = createSessionActivityIntent(arguments)
+            mediaSession = navigator.session(applicationContext, arguments.bookId.toString(), activityIntent)
+                .also { addSession(it) }
+            mediaNavigator = navigator
+            saveLocationJob = navigator.currentLocator
+                .buffer(1, BufferOverflow.DROP_OLDEST)
+                .onEach {  locator ->
+                    delay(3.seconds)
+                    currentBookId?.let { id -> books.saveProgression(locator, id) }
+                }
```

Comment:
I think you can simplify this with a `debounce()`.

But I would try to keep the test app code as simple/dumb as possible and let apps decide if they need to do some optimisations. Also I guess the frequency can be customized with `Configuration.positionRefreshRate`?

```suggestion
                .debounce(3.seconds)
                .onEach {  locator ->
                    currentBookId?.let { id -> books.saveProgression(locator, id) }
                }
```

Reply:
No, `debounce` curiously didn't work here. My guess is that since all events are close enough each from other, they are all skipped.
`positionRefreshRate` should be kept high in order to refresh the time bar. Here we don't need such a high frequency. I think illustrating good practices is not bad.

Reply:
I see, we would need `throttle` but it's not (yet) part of the coroutines library unfortunately. Let's keep your solution until it's available.

Reply:
`sample` should work though.  Replace `debounce` with `sample` in @mickael-menu's suggested change.

Reply:
Will try, thanks!

Reply:
Nice, I didn't know about `sample`. Looks like a `throttleLast`.